dev/PushNotifications/PushManager.cpp

```diff
@@ -4,11 +4,111 @@
 #include "pch.h"
 #include "PushManager.h"
 #include "PushManager.g.cpp"
+#include <winrt\windows.networking.pushnotifications.h>
+#include <winerror.h>
+#include <algorithm>
+#include <list>
+#include <iostream>
+#include <ChannelResult.h>
+#include <winerror.h>
+
+using namespace winrt::Windows::Networking::PushNotifications;
+using namespace winrt;
+
+// MAX BACKOFF TIME IS 16MINUTES 
+#define MAX_BACKOFF_SECONDS 960
+
+// Error code - platform is not connected
+#define WNP_E_NOT_CONNECTED 0x880403E8L
+
+// Error code - platform reconnecting with wns
+#define WNP_E_RECONNECTING  0x880403E9L
+
+// Error code - Current bind operation failed because another bind operation in progress
+#define WNP_E_BIND_USER_BUSY 0x880403FEL
 
 namespace winrt::Microsoft::ProjectReunion::implementation
 {
+    std::list<winrt::guid> PushManager::remoteIdList;
+    std::mutex PushManager::mtx;
+    std::unique_lock<std::mutex> PushManager::lock (mtx, std::defer_lock); // Avoid locking during the constructor call
+
+    bool PushManager::isChannelRequestRetryable(const winrt::hresult& hrException)
+    {
+        switch (hrException)
+        {
+        case HRESULT_FROM_WIN32(ERROR_TIMEOUT):
+            // E_TIMEOUT
+        case WNP_E_NOT_CONNECTED:
+        case WPN_E_OUTSTANDING_CHANNEL_REQUEST:
+        case WNP_E_RECONNECTING
+        case WNP_E_BIND_USER_BUSY:
+            return true;
+        default:
+            return false;
+        }
+    }
+
     Windows::Foundation::IAsyncOperationWithProgress<Microsoft::ProjectReunion::ChannelResult, Microsoft::ProjectReunion::ChannelResult> PushManager::CreateChannelAsync(winrt::guid remoteId)
     {
-        throw hresult_not_implemented();
+        winrt::Microsoft::ProjectReunion::ChannelResult channelResult{ nullptr };
+        auto progress{ co_await winrt::get_progress_token() };
+
+       GUID remoteIdAbiGuid = reinterpret_cast<GUID&>(remoteId);
+       if (remoteIdAbiGuid == GUID_NULL)
+        {
+            channelResult = winrt::make<winrt::Microsoft::ProjectReunion::implementation::ChannelResult>(nullptr, E_INVALIDARG, ChannelStatus::CompletedFailure);
+            co_return channelResult;
+        }
+
+       lock.lock();
+       if (std::find(remoteIdList.begin(), remoteIdList.end(), remoteId) != remoteIdList.end())
```

Comment:
Nit:Only remoteIdList needs to be protected by a lock here.

Reply:
yeah that is what is being done here right? Am I missing something?

Reply:
See my other comment : )